driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -370,14 +370,17 @@ public void onFailure(Throwable t) {
                         if (t instanceof UnsupportedProtocolVersionException) {
                             cluster.manager.logUnsupportedVersionProtocol(host, ((UnsupportedProtocolVersionException) t).getUnsupportedVersion());
                             cluster.manager.triggerOnDown(host, false);
+                            future.set(false);
                         } else if (t instanceof ClusterNameMismatchException) {
                             ClusterNameMismatchException e = (ClusterNameMismatchException) t;
                             cluster.manager.logClusterNameMismatch(host, e.expectedClusterName, e.actualClusterName);
                             cluster.manager.triggerOnDown(host, false);
+                            future.set(false);
                         } else {
-                            logger.warn("Error creating pool to " + host, t);
+                            logger.error("Error creating pool to " + host, t);
```

Comment:
Correct, and the pool creation future would complete normally (although with `false` instead of `true`).

I suppose it was so by design, on the assumption that the failure is transient and the failed host/pool would eventually get created later, so no need to stop the whole session initialization process. 

The change I made here indeed modifies this behavior. My main intent was to account for `AuthenticationException` because when it happens, it usually happens for all hosts, and the session gets created with no pools at all and the user gets a cryptic `BusyConnectionException` when he tries to use the session. 

But you are right concerning OOME: it was being swollen here as well.